Resource monitor

t-davidbrown-msft opened the pull request:
- add cpu monitor returns a float between 0 and 1
- add memory monitor currently return a double between 0 and 1 (this is the current amount of memory a process is using divided by the total amount of memory on the host)
- add tests for cpu monitor and memory monitor. Tests check for 0, less than 1, and null

Commit:
create cpu monitor

- create method for monitoring CPU usage that works for all frameworks using System.Diagnostics.Process
- create one test for ensuring that the monitor works

Commit:
Merge branch 'main' into cpu-monitor
Merging the latest changes into my feature branch

Commit:
update cpu monitor logic

- add CpuMonitor interface
- add CpuMonitor tests
- change CurrentProcessorTime to be the current time on the processor minute the previous

Commit:
finished baseline cpu tuner

- removed unnecessary refactor
- add SimulateCpuLoad to test suite to simulate loading the CPU
- add CoreCount Property to replace function calls to Environment.ProcessorCount

Commit:
add testing

- readd default constructor
- add tests for monitoring interval that is below 100 milliseconds

Commit:
create memory monitor

- create processor directive to create memory monitor for netstandard2.0 or greater, but this is not a great implementation
- create method for MonitorMemoryUsage is otherwise good
- create tests to check whether it is below 1, above 0, and not null

Commit:
add memory check for netstandard20

Commit:
finish cpu and memory monitors

- add cpu monitor returns a float between 0 and 1
- add memory monitor currently return a double between 0 and 1 (this is the current amount of memory a process is using divided by the total amount of memory on the host)
- add tests for cpu monitor and memory monitor. Tests check for 0, less than 1, and null

t-davidbrown-msft commented:
@microsoft-github-policy-service agree company="Microsoft"



Respectfully,





David Brown

Software Engineer – Microsoft LEAPer

Azure Storage Developer Experience | Azure Core/Azure Storage

***@***.******@***.***>



[MSFT_logo_Gray DE sized SIG1.png]



________________________________
From: microsoft-github-policy-service[bot] ***@***.***>
Sent: Saturday, March 1, 2025 9:23 AM
To: Azure/azure-sdk-for-net ***@***.***>
Cc: David Brown ***@***.***>; Mention ***@***.***>
Subject: Re: [Azure/azure-sdk-for-net] Resource monitor (PR #48501)


@t-davidbrown-msft<https://github.com/t-davidbrown-msft> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]


Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree


  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"


Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<http://www.opensource.org/>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

—
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-sdk-for-net/pull/48501#issuecomment-2692257212>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BPVTLLHIL2UILNOBS3OVXDD2SG7G3AVCNFSM6AAAAABYEES33OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSGI2TOMRRGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[microsoft-github-policy-service[bot]]microsoft-github-policy-service[bot] left a comment (Azure/azure-sdk-for-net#48501)<https://github.com/Azure/azure-sdk-for-net/pull/48501#issuecomment-2692257212>

@t-davidbrown-msft<https://github.com/t-davidbrown-msft> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]


Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree


  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"


Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<http://www.opensource.org/>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

—
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-sdk-for-net/pull/48501#issuecomment-2692257212>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BPVTLLHIL2UILNOBS3OVXDD2SG7G3AVCNFSM6AAAAABYEES33OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSGI2TOMRRGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.DataMovement](https://apiview.dev/Assemblies/Review/fffb907bd4224ac4b0f95dfb164a72e3?revisionId=638682becb4c4147920840f1f12b815f)

Commit:
Merge branch 'main' into resource-moni o
merge main into feature branch

Commit:
rename CpuMonitor to ResourceMonitor

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ResourceMonitor.cs:
```diff
@@ -134,6 +167,36 @@ internal async Task CalibrateComputation(CancellationToken cancellationToken)
+                throw new InvalidOperationException();
+
+            IsRunning = false;
+            _cancellationTokenSource?.Cancel();
```
Ignore this one.

amnguye commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ResourceMonitor.cs:
```diff
@@ -134,6 +167,36 @@ internal async Task CalibrateComputation(CancellationToken cancellationToken)
 
+        public void StartMonitoring()
+        {
+            if (IsRunning)
+                return;
+
+            IsRunning = true;
```
Should this be behind a lock / semaphoreslim ?

jalauzon-msft replied:
Setting a boolean is atomic: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#96-atomicity-of-variable-references

Also I'm not sure we'll be using this from multiple threads since there will only be one per transfer manager and I think the transfer manager will be the one starting/stopping it.

amnguye replied:
Hm then that implementation of how transfer manager decides to start and stop it will be interesting, since it will have to juggle and determine from other jobs when to start and stop, when multiple jobs are running at the same time. But yes I agree probably a later problem

amnguye commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ResourceMonitor.cs:
```diff
@@ -156,5 +219,87 @@ private static async Task MonitoringWorker(
+            // Process.TotalProcessorTime.TotlMilliseconds returns the processing time across all processors
+            // Environment.ProcessorCount returns then total number of cores (which includes physical cores plus hyper
+            // threaded cores
+            return (float)(CurrentProcessorTime - PreviousProcessorTime) / (float)(_monitoringInterval.TotalMilliseconds * CoreCount);
```
Not sure if this is necessary but should we look out for accidentally dividing by 0?

t-davidbrown-msft replied:
I just set a default `_monitoringInterval` to 1 second. But I think you're right; let me add something to ensure that we can't do like Chuck Norris 😏: Divide by 0. 🤣

t-davidbrown-msft commented:
> Hm so I'm not sure which commit did this, but maybe when you attempted to merge main into your branch, it looks like it didn't give. and so now it looks like you changed 1.4k files when you didn't

May

t-davidbrown-msft commented:
> Hm so I'm not sure which commit did this, but maybe when you attempted to merge main into your branch, it looks like it didn't give. and so now it looks like you changed 1.4k files when you didn't

It looks like it happened when I did a `revert`. Let me fix this.

Commit:
fix merge issue and add changes

- remove interface
- change private properties to fields
- group private fields and properties
- remove public default constrcutor
- combine CPU and memory into one task and run on single thread
- switch order in resource monitor: take reading first, then sleep
- memory monitor returns Bytes instead of percentage
- remove all functionality for getting total memory

Commit:
change interface to concrete class

Commit:
update tests

Commit:
Merge branch 'main' into resource-monitor

Commit:
Merge branch 'main' into resource-monitor

Commit:
removed tests and updated resource monitor

Commit:
increase simulate load time for mac

Commit:
add functionality for mac os
